keps/sig-scheduling/3101-pod-to-node-assignment/README.md

```diff
@@ -0,0 +1,911 @@
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+A user wants a Pod to be scheduled alone on a node to avoid them from contending
+on resources (e.g. claiming all CPU cores available on a node). In other words,
+resource intensive workloads are needed to be scheduled on separate nodes.
+
+Example: As a user, I want to deploy a Spark or MPI job such that only a single
+worker is running on each pod excluding daemonsets.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+Users might be able to circumvent 1:1 pod to node assignments by creating a
+daemonset with node affinity to a specific node.
+
+#### Possible Mitigations
+
+-   Cluster admins can use one of the following to prevent users from creating
+    daemonsets
+    -   [ResourceQuota](https://kubernetes.io/docs/concepts/policy/resource-quotas/)
+        -   Object count quota (`count/daemonsets`) can be used for this
+            purpose.
+    -   kubernetes
+        [RBAC](https://kubernetes.io/docs/concepts/policy/resource-quotas/) to
+        restrict daemonsets creation to admins/specific users.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The idea is to create a scheduler filter plugin such that given a pod and a
+node, a node is filtered out (as in considered no fit for the pod) if any of the
+following is true (assume the API is the annotation k8s.io/pod-per-node=true):
+
+-   The incoming pod is not a daemonset pod or static pod (by checking
+    OwnerReference) and there is a pod with k8s.io/pod-per-node=true already
+    assigned to the node,
+-   If the incoming pod has the k8s.io/pod-per-node=true annotation and the node
+    hosts at least one non daemonset pod or a static pod.
+-   The above logic would also be added to the set of predicates that kubelet
+    tests on pod admission. This will prevent users from working around the
+    scheduler to avoid 1:1 enforcement.
+
+The annotation is not a proper API, so the next section proposes what the API
+should look like in the pod spec.
+
+### Introduce a new API Field
+
+A new API field `pod.spec.podsPerNodeAssignmentPolicy` could be introduced to be
```

Comment:
I am thinking that this name and type are fairly long, perhaps `CoexistPolicy` is another option. Coexist is kind of a "toleration", which is what this feature is about, a special type of toleration, and the semantics is: what other pods a pod can coexist with on the same node. And so `Many` -> `Any`, and `SingleExcludingDaemonsetAndStaticPods` -> `DaemonsetAndStaticPods`

Reply:
yes, `CoexistPolicy` sounds much better. Updated.